allennlp/commands/train.py

```diff
@@ -373,19 +203,27 @@ def train_model(params: Params,
     logger.info("Loading the best epoch weights.")
     best_model_state_path = os.path.join(serialization_dir, 'best.th')
     best_model_state = torch.load(best_model_state_path)
-    best_model = model
+    best_model = trainer.model
     best_model.load_state_dict(best_model_state)
 
-    if test_data and evaluate_on_test:
+    # Hack to get the test dataset loaded for evaluation, since now only
+    # Trainer.from_params knows about it.
+    if evaluate_on_test and "test_data_path" in orig_params:
```

Comment:
Perhaps this is unavoidable, but it's a bit unfortunate for a refactor PR to introduce this hack. Seems like this use-case could help us guide the API design.

Reply:
Ah, right, I was going to comment on this, but forgot - why not just put the test evaluation into the `train()` function itself?  We already do validation there, why not have the last piece of `train` also do test, if you asked for it?  That would simplify a lot of this, I think.

Reply:
I have two conflicting goals here:

(1) I don't want all of the datasets / iterators stuff to be known to `commands.train`, because I don't want to insist that a trainer necessarily has any of that stuff
(2) I want to somehow preserve the current `evaluate_on_test` behavior

please let me know if you can think of a better way to accomplish these

Reply:
this works, the only downside is that now "load best weights" has to happen twice, because there's no obvious way to return them (you could make that *always* happen and then `commands.train` just has to assume it's the case, but that seems dangerous)

Reply:
It seems reasonable to me to have it always happen; we just document that `SingleTaskTrainer.train()` returns a model with the best weights loaded.  That was the assumed behavior already, I thought.